pkg/gui/controllers/branches_controller.go

```diff
@@ -97,9 +97,75 @@ func (self *BranchesController) GetKeybindings(opts types.KeybindingsOpts) []*ty
 			Handler:     self.checkSelectedAndReal(self.rename),
 			Description: self.c.Tr.LcRenameBranch,
 		},
+		{
+			Key:         opts.GetKey(opts.Config.Branches.SetUpstream),
+			Handler:     self.checkSelected(self.setUpstream),
+			Description: "(Un)set upstream",
+			OpensMenu:   true,
+		},
 	}
 }
 
+// TODO: this is the same as SyncController.parseUpstream
+func (self *BranchesController) parseUpstream(upstream string) (string, string, error) {
+	var upstreamBranch, upstreamRemote string
+	split := strings.Split(upstream, " ")
+	if len(split) != 2 {
+		return "", "", errors.New(self.c.Tr.InvalidUpstream)
+	}
+
+	upstreamRemote = split[0]
+	upstreamBranch = split[1]
+
+	return upstreamRemote, upstreamBranch, nil
+}
+
+func (self *BranchesController) promptForUpstream(currentBranch *models.Branch, onConfirm func(string) error) error {
+	// TODO: how to get this from `gui`?
+	// suggestedRemote := self.getSuggestedRemote()
+	suggestedRemote := "origin"
+
+	return self.c.Prompt(types.PromptOpts{
+		Title:               self.c.Tr.EnterUpstream,
+		InitialContent:      suggestedRemote + " " + currentBranch.Name,
+		FindSuggestionsFunc: self.helpers.Suggestions.GetRemoteBranchesSuggestionsFunc(" "),
+		HandleConfirm:       onConfirm,
+	})
+}
+
+func (self *BranchesController) setUpstream(selectedBranch *models.Branch) error {
+	return self.c.Menu(types.CreateMenuOptions{
+		Title: "(Un)set upstream",
```

Comment:
Conventionally we would say `Set/Unset upstream` (if memory serves correct). We'll also want to i18n'ise this

Reply:
Sure, I'll deal with that as soon as we iron out the kinks regarding the logic.